### ISSUE TITLE
Namespace binding applied inconsistently

### ISSUE THREAD

**Issue by mapfde** [RDFLib/rdflib#1019]
Hi,

I have a small graph with data from dbpedia in file dbpedia_anime_2.ttl (appended as TXT). I want to simplify it by using an additional namespace for dbpedia ressources. I use the following code:

```g2 = Graph()
dbp_ns = Namespace('http://dbpedia.org/resource/')
g2.namespace_manager.bind('dbpres', dbp_ns, replace=True, override=True)

g2.parse("dbpedia_anime_2.ttl", format="turtle")

fh = open ('dbpedia_anime_3.ttl', 'wb')
fh.write(g2.serialize(format="turtle"))
fh.close()
```
which runs fine and creates the file dbpedia_anime_3.ttl (also appended as TXT). 

The problem is, that the new namespace is sometimes applied to an URI and sometimes not. I was expecting RDFlib to consistently shorten all applicable URI using the new namespace.

Am I doing something wrong?

[dbpedia_anime_2.ttl.TXT](https://github.com/RDFLib/rdflib/files/4523416/dbpedia_anime_2.ttl.TXT)
[dbpedia_anime_3.ttl.TXT](https://github.com/RDFLib/rdflib/files/4523417/dbpedia_anime_3.ttl.TXT)

Kind regards,
Magnus

**Comment by tgbugs** [comment]
This is happening because the suffixes after `/resource/` have characters e.g. `(` `)` that cannot be present in a curie suffix so rdflib won't shorten them.

**Comment by mapfde** [comment]
That would explain some, but not many of the examples. What is wrong with
````
<http://dbpedia.org/resource/1974_in_Japanese_television>
<http://dbpedia.org/resource/22nd_century_in_fiction>
<http://dbpedia.org/resource/2014_in_Japan>
```
starting with a number should not be problematic. Or is it?

**Comment by mapfde** [comment]
This leaves the question open, *why* `(` and `)` are a problem. They are considered "safe" characters in URIs according to https://www.ietf.org/rfc/rfc1738.txt. They are also part of the ABNF for IRIs (https://www.ietf.org/rfc/rfc3987.txt) which is referenced by the Curie def (https://www.w3.org/TR/curie/)

Am I missing something?

**Comment by tgbugs** [comment]
Starting with a number should have been fixed in 5.0.0 (and those examples do shorten as expected for me).

You aren't missing anything in the spec, just a bit of rdflib history.

The short version is that curies != qnames and rdflib uses the qname rules because it assumes worst case that you will serialize to rdf/xml which must use qnames. Figuring out how to support both curies and qnames in the internal namespace manager is what will be required to fix this. See https://github.com/RDFLib/rdflib/pull/649 for more. 

This is indeed a bug in rdflib which stems from conflation of curies with qnames. As you point out https://www.ietf.org/rfc/rfc3987.txt allows them, however https://www.w3.org/TR/1999/REC-xml-names-19990114/#NT-NCName does not. I made it part way through deconflating the two, however there are still many places in the codebase where there are deeply embedded assumptions that curies == qnames. The reason this is an issues is because the prefix/namespace split is deep in the internals of in the rdflib implementation, not just at the surface (and needs to be for performance reasons). Therefore at the moment any prefix that is computed by rdflib follows the qname rules, not the curie rules. Curies inherited from another source can violate this (which prevents conversion of some ttl files to rdf/xml because there is no valid qname for some iris that are used as predicates).
